Mathlib/Topology/ApproximateUnit.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2024 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import Mathlib.Topology.MetricSpace.Bounded
+import Mathlib.Topology.Algebra.Monoid
+
+/-! # Approximate units
+
+An *approximate unit* is a filter `l` such that multiplication on the left (or right) by `m : α`
+tends to `𝓝 m` along the filter, and additionally `l ≠ ⊥` and `Disjoint l (cobounded α)`.
+
+Examples of approximate units include:
+
+- The trivial approximate unit `pure 1` in a normed ring.
+- `𝓝 1` or `𝓝[≠] 1` in a normed ring (note that the latter is disjoint from `pure 1`).
+- In a C⋆-algebra, the filter generated by the sections `fun a ↦ {x | a ≤ x} ∩ closedBall 0 1`,
+  where `a` ranges over the positive elements of norm strictly less than 1.
+-/
+
+open Filter Topology Bornology
+
+/-- An *approximate unit* is a proper bounded filter (i.e., `≠ ⊥` and disjoint from `cobounded α`)
+such that multiplication on the left (and separately on the right) by `m : α` tends to `𝓝 m` along
+the filter. -/
+structure Filter.IsApproximateUnit {α : Type*} [TopologicalSpace α] [Mul α] [Bornology α]
+    (l : Filter α) : Prop where
+  /-- Multiplication on the left by `m` tends to `𝓝 m` along the filter. -/
+  tendsto_mul_left m : Tendsto (m * ·) l (𝓝 m)
+  /-- Multiplication on the right by `m` tends to `𝓝 m` along the filter. -/
+  tendsto_mul_right m : Tendsto (· * m) l (𝓝 m)
+  /-- The filter is bounded. -/
+  disjoint_cobounded : Disjoint l (cobounded α)
```

Comment:
Do you have a reference for this assumption ? Naively it feels a bit out of place (as in, we're going to introduce a stronger assumption in the normed setting anyway), but I'm guessing this has something to do with matrices ?

Reply:
No, it's not about matrices (they don't have a bornology either). It was more just that when I went reading, it seemed that unbounded approximate identities were quite pathological (e.g., see [here](https://www.cambridge.org/core/services/aop-cambridge-core/content/view/43C36850709EE1DF6E22AC965C11722F/S0017089500008703a.pdf/unbounded-approximate-identities-in-normed-algebras.pdf), although I think I had found even more convincing evidence before, but I don't recall what it was).

However, I'm not committed to keeping this condition if we think it's better to remove it.

Reply:
Perhaps we could remove the condition for now, and add it back later if not having it ends up being annoying.

Reply:
I've removed it, but if we prefer to put it back (even in this PR) it's just a `git revert` away.

Reply:
Indeed my thinking was not that we should try to work hard to include unbounded approximate identities (they seem pathological intuitively, and I'm glad I now know more precisely what goes wrong). My thinking was that you were going to introduce a more specific "supported on the closed unit ball" class for normed algebras anyway in #18506 (this is the general definition Bourbaki uses in a general normed algebra). Now, I don't know why I thought that because that's not what #18506 is. More interestingly, I don't know (let's say in normed algebras) wether there is any crucial distinction between "bounded approximate units" and "approximate units in the closed unit ball", the distinction being that we can talk about the former but not the latter if we just have a bornology.

So really I still don't know what is the right thing to do here. I think by default I would prefer to not add the condition until we're sure it's not going to backfire (e.g if someone wants to write down an approximate unit in the algebra of compactly supported smooth functions and convolution, I don't want them to have to define a bornology and everything), but if at any points that causes trouble I'm completely fine with adding it back.

Note: adding the bornology to matrices would be perfectly fine, since it's completely canonical

Reply:
> Now, I don't know why I thought that because that's not what #18506 is.

Well, it *does* include that `IncreasingApproximateUnit`, but that also has the condition that the elements are nonnegative, so there is a potential in-between predicate here.

> any crucial distinction between "bounded approximate units" and "approximate units in the closed unit ball"

This distinction matters for C⋆-algebras in the context of `IncreasingApproximateUnit`s, because the point is that if `0 ≤ x` and `‖x‖ ≤ 1`, then `0 ≤ 1 - x`, which is used *everywhere*.

> Note: adding the bornology to matrices would be perfectly fine, since it's completely canonical

Of course, but we don't have it at the moment I don't think.